carma-messenger-core/cpp_message/src/cpp_message.cpp

```diff
@@ -0,0 +1,181 @@
+{
+
+    void Message::initialize()
+    {
+        nh_.reset(new ros::CARMANodeHandle());
+        pnh_.reset(new ros::CARMANodeHandle("~"));
+        // initialize pub/sub
+        outbound_binary_message_pub_ = nh_->advertise<cav_msgs::ByteArray>("outbound_binary_msg", 5);
+        inbound_binary_message_sub_ = nh_->subscribe("inbound_binary_msg", 5, &Message::inbound_binary_callback, this);
+        outbound_geofence_request_message_sub_ = nh_->subscribe("outbound_control_request", 5, &Message::outbound_control_request_callback, this);
+        inbound_geofence_request_message_pub_ = nh_->advertise<j2735_msgs::ControlRequest>("inbound_control_request", 5);
+    }
+
+    void Message::inbound_binary_callback(const cav_msgs::ByteArrayConstPtr& msg)
+    {
+        // only handle ControlRequest for now
+        if(msg->messageType == "ControlRequest") {
+            j2735_msgs::ControlRequest output;
+            // decode results
+            asn_dec_rval_t rval;
+	        MessageFrame_t* message = 0;
+            // copy from vector to array
+	        std::vector<uint8_t> array = msg->content;
+            auto len = array.size();
+            uint8_t buf[len];
+            for(auto i = 0; i < len; i++) {
+	            buf[i] = array[i];
+	        }
+            // use asn1c lib to decode
+	        rval = uper_decode(0, &asn_DEF_MessageFrame, (void **) &message, buf, len, 0, 0);
+
+            // if decode successed
+            if(rval.code == RC_OK) {
+                // convert version from char array to string
+                std::string version;
+                auto str_len = message->value.choice.TestMessage04.body.version.size;
+                for(auto i = 0; i < str_len; i++)
+                {
+                    version += message->value.choice.TestMessage04.body.version.buf[i];
+                }
+                output.version = version;
+                // copy scale field
+                output.scale = message->value.choice.TestMessage04.body.scale;
+                // copy bounds
+                auto bounds_count = message->value.choice.TestMessage04.body.bounds.list.size;
+                for(auto i = 0; i < bounds_count; i++) {
+                    j2735_msgs::ControlBounds bounds;
+                    // copy lat/lon
+                    bounds.latitude = message->value.choice.TestMessage04.body.bounds.list.array[i]->lat;
+                    bounds.longitude = message->value.choice.TestMessage04.body.bounds.list.array[i]->lon;
+                    // copy offset array to boost vector
+                    auto count = message->value.choice.TestMessage04.body.bounds.list.array[i]->offsets.list.size;
+                    for(auto j = 0; j < count; j++) {
+                        bounds.offsets[j] = *message->value.choice.TestMessage04.body.bounds.list.array[i]->offsets.list.array[j];
+                    }
+                    // recover a long value from 8-bit array
+                    long temp = 0;
```

Comment:
might be better to use the bit specified type here, for clarity.
Like int64_t or which ever applied (uint64_t etc.)